#### Knowledge Sharing Content
# <center> One Liners - Data Science
#### [Bhanu Pratap Singh](https://www.linkedin.com/in/bpst/)

## Two Dimensional Array Arithmetic

In [1]:
# Importing Packages
import numpy as np

c:\program files\python37\lib\site-packages\numpy\_distributor_init.py:32: UserWarning: loaded more than 1 DLL from .libs:
c:\program files\python37\lib\site-packages\numpy\.libs\libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll
c:\program files\python37\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
  stacklevel=1)


In [2]:
# Creating a 1D array from a list
a = np.array([1, 2, 3])
print(a)

[1 2 3]


In [3]:
# Creating a 2D array from a list of lists
b = np.array([[1, 2],
              [3, 4]])
print(b)

[[1 2]
 [3 4]]


In [4]:
# Creating a 3D array from a list of lists of lists
c = np.array([[[1, 2], [3, 4]],
              [[5, 6], [7, 8]]])
print(c)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


__Note__

* A one-dimensional array corresponds to a simple `list` of numerical values (in fact, NumPy arrays can contain other data types too, but we’ll focus on numbers here). 
* A two-dimensional array corresponds to a nested `list of lists` of numerical values. 
* A three-dimensional array corresponds to a nested `list of lists of lists` of numerical values. 

The number of opening and closing brackets gives us the dimensionality of the NumPy array.

In [6]:
# First Array
a = np.array([[1, 0, 0],
              [1, 1, 1],
              [2, 0, 0]])

# Second Array
b = np.array([[1, 1, 1],
              [1, 1, 2],
              [1, 1, 2]])

In [7]:
# Addition
print(a + b)

[[2 1 1]
 [2 2 3]
 [3 1 2]]


In [8]:
# Substraction
print(a - b)

[[ 0 -1 -1]
 [ 0  0 -1]
 [ 1 -1 -2]]


In [9]:
# Multiplication
print(a * b)

[[1 0 0]
 [1 1 2]
 [2 0 0]]


In [10]:
# Division
print(a / b)

[[1.  0.  0. ]
 [1.  1.  0.5]
 [2.  0.  0. ]]


In [11]:
# Division
print(a // b)

[[1 0 0]
 [1 1 0]
 [2 0 0]]


NumPy provides a lot more capabilities for manipulating arrays, including the `np.max()` function, which calculates the maximum value of all values in a NumPy array. The `np.min()` function calculates the minimum value of all values in a NumPy array. The `np.average()` function calculates the average value of all values in a NumPy array.

In [12]:
# max function
print(np.max(a))

2


In [13]:
# min function
print(np.min(a))

0


In [14]:
# average function
print(np.average(a))

0.6666666666666666


The maximum value of all values in the NumPy array is 2, the minimum value is 0, and the average is (1 + 0 + 0 + 1 + 1 + 1 + 2 + 0 + 0) / 9 = 2/3.

### Promblem Statement - I

Find the maximum after-tax income in a group of people, given their yearly salary and tax rates.

#### Solution

Let’s tackle this problem by using the salary data of Alice, Bob, and Tim. 

| Name | 2018 | 2019 | 2020 |
| --- | --- | --- | --- |
| Alice | 99 | 101 | 103 |
| Bob | 110 | 108 | 105 |
| Tim | 90 | 88 | 85 |

<i>**Note**: Numbers are yearly salary in ($1000)</i>

It seems like Bob has enjoyed the highest salary in the last three years. But is he actually bringing home the most money, considering the individual tax rates of our three friends?

In [18]:
# Data: yearly salary in ($1000) [2017, 2018, 2019]
alice = [99, 101, 103]
bob = [110, 108, 105]
tim = [90, 88, 85]

salaries = np.array([alice, bob, tim])
taxation = np.array([[0.2, 0.25, 0.22],
                     [0.4, 0.5, 0.5],
                     [0.1, 0.2, 0.1]])

# One-liner
max_income = np.max(salaries - salaries * taxation)

# Result
print(max_income)

81.0


Wow, amazing however how it works!!!

Let's see that now,

In [17]:
# printing salaries
salaries

array([[ 99, 101, 103],
       [110, 108, 105],
       [ 90,  88,  85]])

We put the data into a two-dimensional NumPy array with three rows (one row for each person: Alice, Bob, and Tim) and three columns (one column for each year: 2017, 2018, and 2019).

Basically, now we have two two-dimensional arrays: `salaries` holds the yearly incomes, and `taxation` holds the taxation rates for each person and year.

Let's analyse income computation,

**max_income = np.max(salaries - salaries * taxation)**

To calculate the after-tax income, you need to deduct the tax (as a dollar amount) from the gross income stored in the array salaries. For this, we use the overloaded NumPy operators - and *, which perform element-wise computations on the NumPy arrays.

The element-wise multiplication of two multidimensional arrays is called the `Hadamard product`.

In [22]:
# salaries for all for all years
print(salaries - salaries * taxation)

[[79.2  75.75 80.34]
 [66.   54.   52.5 ]
 [81.   70.4  76.5 ]]


Here, we can see that Bob’s large income is significantly reduced after paying 40 percent and 50 percent tax rates, shown in the second row.

In [23]:
# printing max income
max_income

81.0

The `np.max()` function simply finds the maximum value in the array, which we store in max_income. 

Thus, the maximum value is Tim’s $90,000 income in 2017, which is taxed at only 10 percent — the result of the one-liner is 81. (again, the dot indicates the float data type).

### Problem Statement - II
We have data for a variety of professions, and we want to increase the salaries of just the data scientists by 10 percent every other year.

| Professions | 2019 | 2020 | 2021 |
| --- | --- | --- | --- |
| Data Scientist  | 130 | 132 | 137 |
| Product Manager  | 127 | 140 | 145 |
| Designer  | 118 | 118 | 127 |
| Software Engineer   | 129 | 131 | 137 |

<i>**Note**: Numbers are yearly salary in ($1000)</i>

In [24]:
# Data: yearly salary in ($1000) [2019, 2020, 2021]
dataScientist =     [130, 132, 137]
productManager =    [127, 140, 145]
designer =          [118, 118, 127]
softwareEngineer =  [129, 131, 137]

employees = np.array([dataScientist,
                      productManager,
                      designer,
                      softwareEngineer])

# One-liner
employees[0,::2] = employees[0,::2] * 1.1

# Result
print(employees)

[[143 132 150]
 [127 140 145]
 [118 118 127]
 [129 131 137]]


In [43]:
# year 2019
round(130 * 1.1)

143

In [42]:
# year 2021
round(137 * 1.1)

151

__Result__

Notice the bold values in output,

[[**143** 132 **150**]<br>
 [127 140 145]<br>
 [118 118 127]<br>
 [129 131 137]]

Salaries of only data scientist are increased 10% every other year remaining salaries are same as earlier.

### Problem Statement - III

Find cities with above-average pollution peaks from a set of data.

| Cities | Day 1 | Day 2 | Day 3 | Day 4 | Day 5 | Day 6 |
| --- | --- | --- | --- | --- | --- | --- |
| Hong Kong  | 42 | 40 | 41 | 43 | 44 | 43 |
| New York  | 30 | 31 | 29 | 29 | 29 | 30 |
| Berlin  | 8 | 13 | 31 | 11 | 11 | 9 |
| Montreal   | 11 | 11 | 12 | 13 | 11 | 12 |

<i>**Note**: Numbers are daily air quality index mesurements in some relvent units</i>

In [26]:
# Data: air quality index AQI data (row = city)
X = np.array(
    [[ 42, 40, 41, 43, 44, 43 ],  # Hong Kong
     [ 30, 31, 29, 29, 29, 30 ],  # New York
     [ 8, 13, 31, 11, 11, 9 ],    # Berlin
     [ 11, 11, 12, 13, 11, 12 ]]) # Montreal

cities = np.array(["Hong Kong", "New York", "Berlin", "Montreal"])

# One-liner
polluted = set(cities[np.nonzero(X > np.average(X))[0]])

# Result
print(polluted)

{'Hong Kong', 'Berlin', 'New York'}


### Problem Statement - IV
Find the names of the Instagram superstars with more than 100 million followers!

| Instagram Influencer | Followers  |
| --- | --- |
| @instagram | 232 |
| @selenagomez | 133 |
| @victoriassecret | 59 |
| @cristiano | 120 |
| @beyonce | 111 |
| @nike | 76 |

<i>**Note**: Numbers are number of followers (in millions) for instagram influencer</i>

In [28]:
# Data: popular Instagram accounts (millions followers)
inst = np.array([[232, "@instagram"],
                 [133, "@selenagomez"],
                 [59,  "@victoriassecret"],
                 [120, "@cristiano"],
                 [111, "@beyonce"],
                 [76,  "@nike"]])

# One-liner
superstars = inst[inst[:,0].astype(float) > 100, 1]

# Results
print(superstars)

['@instagram' '@selenagomez' '@cristiano' '@beyonce']


__Result__

The output shows influencers from our data set with more than 100 million Instagram followers

### Problem Statement - V
Given an array of temperature values, replace every seventh temperature value with the average of the last seven days (including the seventh day’s temperature value).

| Sensor Reading | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Morning  | 1 | 2 | 3 | 4 | 3 | 4 | 4 |
| Afternoon  | 5 | 3 | 3 | 4 | 3 | 4 | 6 |
| Night  | 6 | 5 | 5 | 5 | 4 | 5 | 5 |

<i>**Note**: Numbers are temperature sensor readings for entire week</i>

In [29]:
# Sensor data (Mo, Tu, We, Th, Fr, Sa, Su)
tmp = np.array([1, 2, 3, 4, 3, 4, 4,
                5, 3, 3, 4, 3, 4, 6,
                6, 5, 5, 5, 4, 5, 5])

# One-liner
tmp[6::7] = np.average(tmp.reshape((-1,7)), axis=1)

# Result
print(tmp)

[1 2 3 4 3 4 3 5 3 3 4 3 4 4 6 5 5 5 4 5 5]


In [37]:
# For Morning
np.average([1, 2, 3, 4, 3, 4, 4])

3.0

In [38]:
# For Afternoon
np.average([5, 3, 3, 4, 3, 4, 6])

4.0

In [39]:
# For Night
np.average([6, 5, 5, 5, 4, 5, 5])

5.0

__Result__

Notice bold values in output, initial values were 4, 6 and 3 respectively. However now they are replaced by average value for week

[1 2 3 4 3 4 **3** 5 3 3 4 3 4 **4** 6 5 5 5 4 5 **5**]

### Problem Statement - VI
Find the names of the top three students with the highest SAT scores. 

| Name | John | Bob | Alice | Joe | Jane | Frank | Carl |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Scores | 1100 | 1256 | 1543 | 1043 | 989 | 1412 | 1343 |

In [30]:
# Data: SAT scores for different students
sat_scores = np.array([1100, 1256, 1543, 1043, 989, 1412, 1343])
students = np.array(["John", "Bob", "Alice", "Joe", "Jane", "Frank", "Carl"])

# One-liner
top_3 = students[np.argsort(sat_scores)][:-4:-1]

# Result
print(top_3)

['Alice' 'Frank' 'Carl']


__Result__

Alice, Frank, and Carl have the highest SAT scores of 1543, 1412, and 1343, respectively.

### Problem Statement - VII
Create a filter function that takes a list of books x and a minimum rating y and returns a list of potential bestsellers that have higher than minimum rating, y'>y.

| Title | Rating  |
| --- | --- |
| Coffee Break NumPy | 4.6 |
| Lord of the Rings | 5.0 |
| Harry Potter | 4.3 |
| Winnie-the-Pooh | 3.9 |
| The Clown of God | 2.2 |
| Coffee Break Python | 4.7 |

In [31]:
# Data (row = [title, rating])
books = np.array([['Coffee Break NumPy', 4.6],
                  ['Lord of the Rings', 5.0],
                  ['Harry Potter', 4.3],
                  ['Winnie-the-Pooh', 3.9],
                  ['The Clown of God', 2.2],
                  ['Coffee Break Python', 4.7]])

# One-liner
predict_bestseller = lambda x, y : x[x[:,1].astype(float) > y]

# Results
print(predict_bestseller(books, 3.9))

[['Coffee Break NumPy' '4.6']
 ['Lord of the Rings' '5.0']
 ['Harry Potter' '4.3']
 ['Coffee Break Python' '4.7']]


### Problem Statement - VIII
Find all outlier days for which the statistics deviate more than the standard deviation from their mean statistics.

| Days | Daily Active Users | Bounce Rate | Average Session Duration (seconds) |
| --- | --- | --- | --- |
| Day 1  | 815 | 70 | 115 |
| Day 5  | 767 | 80 | 50 |
| Day 10 | 912 | 74 | 77 |
| Day 15 | 554 | 88 | 70 |
| Day 20 | 1008 | 65 | 128 |

<i>**Note**: Numbers represents the column record mesurements in their respective units</i>

In [32]:
# Website analytics data:
# (row = day), (col = users, bounce, duration)
a = np.array([[815, 70, 115],
              [767, 80, 50],
              [912, 74, 77],
              [554, 88, 70],
              [1008, 65, 128]])
mean, stdev = np.mean(a, axis=0), np.std(a, axis=0)
# [811.2  76.4  88. ], [152.97764543   6.85857128  29.04479299]

# One-liner
outliers = ((np.abs(a[:,0] - mean[0]) > stdev[0])
            * (np.abs(a[:,1] - mean[1]) > stdev[1])
            * (np.abs(a[:,2] - mean[2]) > stdev[2]))

# Result
print(a[outliers])

[[1008   65  128]]


__Result__

There is only a single row for which all three columns are outliers

### Problem Statement - IX
What fraction of customers bought two ebooks together? Based on this data, the recommender system can offer customers a book “bundle” to buy if it sees that they originally intended to buy a single book.

| Customer  | Course 1 | Course 2 | eBook 1 | eBook 2 |
| --- | --- | --- | --- | --- |
| Customer 1  | 0 | 1 | 1 | 0 | 
| Customer 2  | 0 | 0 | 0 | 1 | 
| Customer 3  | 1 | 1 | 0 | 0 |
| Customer 4  | 0 | 1 | 1 | 1 |
| Customer 5  | 1 | 1 | 1 | 0 |
| Customer 6  | 0 | 1 | 1 | 0 |
| Customer 7  | 1 | 1 | 0 | 1 |
| Customer 8  | 1 | 1 | 1 | 1 |

In [33]:
# Data: row is customer shopping basket
# row = [course 1, course 2, ebook 1, ebook 2]
# value 1 indicates that an item was bought.
basket = np.array([[0, 1, 1, 0],
                   [0, 0, 0, 1],
                   [1, 1, 0, 0],
                   [0, 1, 1, 1],
                   [1, 1, 1, 0],
                   [0, 1, 1, 0],
                   [1, 1, 0, 1],
                   [1, 1, 1, 1]])

# One-liner
copurchases = np.sum(np.all(basket[:,2:], axis = 1)) / basket.shape[0]

# Result
print(copurchases)

0.25


__Result__

25 % customers bought both ebooks together

#### Another Problem
Find the two items that were purchased most often together

In [34]:
# One-liner (broken down in two lines;)
copurchases = [(i,j,np.sum(basket[:,i] + basket[:,j] == 2))
               for i in range(4) for j in range(i+1,4)]

# Result
print(max(copurchases, key=lambda x:x[2]))

(1, 2, 5)


__Result__

The second and third products have been purchased together five times. No other product combination reaches copurchasing power this high. Hence, we can tell our boss to upsell product 2 when selling product 1, and vice versa.